doc/specs/stdlib_selection.md

```diff
@@ -0,0 +1,340 @@
+      ! index >= 2
+      call select(array, k, kth_smallest, left=2)
+      print*, kth_smallest ! print 5.0
+
+      k = 6
+      ! Due to the previous two calls to select, we know for sure this is in
+      ! an index >= 2 and <= 7
+      call select(array, k, kth_smallest, left=2, right=7)
+      print*, kth_smallest ! print 4.0
+
+    end program demo_select
+```
+
+#### `arg_select` - find the kth smallest value in an input array
+
+##### Status
+
+Experimental
+
+##### Description
+
+Returns the index of the k-th smallest value of `array(:)`, and also partially sorts
+the index-array `indx(:)` such that `all(array(indx(1:k)) <= array(indx(k)))`  and
+`all(array(indx(k)) <= array(indx((k+1):size(array))))`
+
+##### Syntax
+
+`call [[stdlib_selection(module):arg_select(interface)]]( array, indx, k, kth_smallest [, left, right ] )`
+
+##### Class
+
+Generic subroutine.
+
+##### Arguments
+
+`array` : shall be a rank one array of any of the types:
+`integer(int8)`, `integer(int16)`, `integer(int32)`, `integer(int64)`,
+`real(sp)`, `real(dp)`, `real(qp)`. It is an `intent(in)` argument. On input it is
+the array in which we search for the kth smallest entry.
+
+`indx`: shall be a rank one array with the same size as `array`, containing all integers
+from `1:size(array)` in any order. It is of any of the types:
+`integer(int8)`, `integer(int16)`, `integer(int32)`, `integer(int64)`. It is an
+`intent(inout)` argument. On return its elements will define a partial sorting of `array(:)` such that:
+ `all( array(indx(1:k-1)) <= array(indx(k)) )` and `all(array(indx(k)) <= array(indx(k+1:size(array))))`.
+
+`k`: shall be a scalar with the same type as `indx`. It is an `intent(in)`
+argument. We search for the `k`-th smallest entry of `array(:)`.
+
+`kth_smallest`: a scalar with the same type as `indx`. It is an `intent(out)` argument,
+and on return it contains the index of the k-th smallest entry of `array(:)`.
+
+`left` (optional): shall be a scalar with the same type as `k`. It is an `intent(in)`
+argument. If specified then we assume the k-th smallest value is definitely contained
+in `array(indx(left:size(array)))`. If not present it is 1. This is typically useful if multiple calls
+to `arg_select` are made, because the partial sorting of `indx` implies constraints on where
+we need to search.
+
+`right` (optional): shall be a scalar with the same type as `k`. It is an `intent(in)`
+argument. If specified then we assume the k-th smallest value is definitely contained
+in `array(indx(1:right))`. If not present it is `size(array)`. This is typically useful if multiple calls
+to `arg_select` are made, because the reordering of `indx` implies constraints on
+where we need to search.
+
+##### Notes
+
+`arg_select` does not modify `array`, unlike `select`.
+
+While it is essential that that `indx` contains the integers `1:size(array)` (in any
+order), the code does not check for this.
```

Comment:
I am not sure to understand what the code doesn't check.

Reply:
clarified in an upcoming commit.
```
While it is essential that that `indx` contains a permutation of the integers `1:size(array)`, 
the code does not check for this. For example if `size(array) == 4`, then we could have 
`indx = [4, 2, 1, 3]` or `indx = [1, 2, 3, 4]`, but not `indx = [2, 1, 2, 4]`. It is the user's 
responsibility to avoid such errors.
```